server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -480,87 +483,125 @@ public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request, Action
-                                            ClusterStateTaskExecutor.unbatched()
-                                        ),
-                                        listener::onFailure
+                    state = blockResult.v1();
+                    final Map<Index, ClusterBlock> blockedIndices = blockResult.v2();
+                    taskContext.success(task.listener.delegateFailure((delegate1, clusterState) -> {
+                        if (blockedIndices.isEmpty()) {
+                            delegate1.onResponse(AddIndexBlockResponse.EMPTY);
+                        } else {
+                            threadPool.executor(ThreadPool.Names.MANAGEMENT)
+                                .execute(
+                                    new WaitForBlocksApplied(
+                                        blockedIndices,
+                                        task.request,
+                                        delegate1.delegateFailure((delegate2, verifyResults) -> {
+                                            clusterService.submitStateUpdateTask(
+                                                "finalize-index-block-["
+                                                    + task.request.getBlock().name
+                                                    + "]-["
+                                                    + blockedIndices.keySet().stream().map(Index::getName).collect(Collectors.joining(", "))
+                                                    + "]",
+                                                new FinalizeBlocksTask(task.request, blockedIndices, verifyResults, delegate2),
+                                                ClusterStateTaskConfig.build(Priority.URGENT),
+                                                finalizeBlocksExecutor
+                                            );
+                                        })
                                     )
-                                )
-                            );
-                    }
+                                );
+                        }
+                    }));
+                } catch (Exception e) {
+                    taskContext.onFailure(e);
                 }
+            }
+
+            return state;
+        }
+    }
+
+    private record AddBlocksTask(AddIndexBlockClusterStateUpdateRequest request, ActionListener<AddIndexBlockResponse> listener)
+        implements
+            ClusterStateTaskListener {
+
+        @Override
+        public void onFailure(Exception e) {
+            listener.onFailure(e);
+        }
 
-                @Override
-                public void onFailure(final Exception e) {
-                    listener.onFailure(e);
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "not called";
+        }
+    }
+
+    private class FinalizeBlocksExecutor implements ClusterStateTaskExecutor<FinalizeBlocksTask> {
```

Comment:
Nit: 

```suggestion
    private static class FinalizeBlocksExecutor implements ClusterStateTaskExecutor<FinalizeBlocksTask> {
```

Reply:
035e71a8b7eb50b276ad62546d7928c9507e56f0, sure, this one doesn't have to not be static (unlike the other Executors in this file, where they do happen to rely on instance fields of the enclosing class).